R/BoostedStacking.R

```diff
@@ -0,0 +1,352 @@
+#'   Needed to exract fix parameters.
+#' @param mm.ps [\code{list containing \link{ParamSet}}]\cr
+#'   Needed to extract fix parameters.
+#' @param x.best [\code{integer(1)}]
+#'   Number of best learners to extract.
+#' @param measure [\code{\link{Measure}}]\cr
+#'   One Measure to apply to "extract from best".
+#' @return [\code{list of x best learners}].
+#' @examples 
+#' \dontrun{
+#' tsk = pid.task
+#' lrns = list(
+#'   makeLearner("classif.gbm", distribution = "bernoulli"),
+#'   makeLearner("classif.randomForest", ntree = 3))
+#' bpt = spt = "prob"
+#' lrns = lapply(lrns, setPredictType, bpt)
+#' mm = makeModelMultiplexer(lrns)
+#' ctrl = makeTuneControlRandom(maxit = 3L)
+#' ps = makeModelMultiplexerParamSet(mm,
+#'   makeIntegerParam("n.trees", lower = 1L, upper = 3L),
+#'   makeIntegerParam("interaction.depth", lower = 1L, upper = 2L),
+#'   makeIntegerParam("ntree", lower = 1L, upper = 3L),
+#'   makeIntegerParam("mtry", lower = 1L, upper = getTaskNFeats(tsk)))
+#' tune.res = tuneParams(learner = mm, task = tsk, resampling = cv2, par.set = ps, control = ctrl)
+#' best = makeXBestLearnersFromMMTuneResult(tune.result = tune.res, 
+#'   model.multiplexer = mm, mm.ps = ps, x.best = 2, measure  =mmce)
+#' best
+#' }
+
+makeXBestLearnersFromMMTuneResult = function(tune.result, model.multiplexer, mm.ps, x.best = 5, measure) {
+  # checks
+  assertClass(tune.result, "TuneResult")
+  assertClass(model.multiplexer, "ModelMultiplexer")
+  assertClass(measure, "Measure")
+  assertInt(x.best, lower = 1)
+  # body
+  measure.name = paste0(measure$id, ".test.mean")
+  opt.grid = as.data.frame(trafoOptPath(tune.result$opt.path), stringsAsFactors = FALSE)
+  #FIXME minimize
+  if (measure$minimize) {
+    ord = order(opt.grid[, measure.name])[1:x.best]
+  } else {
+    ord = rev(order(opt.grid[, measure.name]))[1:x.best]
+  }
+  opt.grid = opt.grid[ord, ]
+  j = sapply(opt.grid, is.factor)
+  opt.grid[j] = lapply(opt.grid[j], as.character)
+  # checks2
+  if(NROW(opt.grid) < x.best) {
+    stopf("'x.best' is %s and cannot be set larger than the number of tuning results in '%s", x.best, quote(tune.result))
+  }
+  #
+  lrns = vector("list", x.best)
+  for (i in 1:nrow(opt.grid)) {
+    # get tuned parameter
+    cl = as.character(opt.grid[i, 1])
+    pars = opt.grid[i, grepl(pattern = cl, names(opt.grid)), drop = FALSE]
+    pars.names.long = names(pars)
+    pars.names = substr(names(pars), nchar(cl) + 2, nchar(names(pars)))
+    names(pars) = pars.names
+    pars.list = as.list(pars)
+    # get and apply trafo
+    #trafo = lapply(pars.names.long, function(x) mm.ps$pars[[x]]$trafo)
+    #trafo = lapply(trafo, function(x) if(is.null(x)) {function(x) x} else x)
+    #names(trafo) = pars.names
+    #pars.list = Map(do.call, trafo, lapply(par.list[names(trafo)], list)) #mapply(do.call, trafo, lapply(par.list[names(trafo)], list))
+    # get fix parameters, and final parameter set
+    pars.list.fix = model.multiplexer$base.learners[[cl]]$par.vals
```

Comment:
when it is possible, we should use the getter functions, e.g. here maybe `getHyperPars` should be better, right?